src/python/pants/source/source_root.py

```diff
@@ -253,6 +254,22 @@ def register_options(cls, register):
 
   @memoized_method
   def get_source_roots(self):
+    deprecated_conditional(
+      lambda: self.options.is_default("unmatched"),
+      removal_version="1.28.0.dev0",
+      entity_description="Defaulting to `--source-unmatched=create`",
+      hint_message=(
+        "`--source-unmatched=create` is an unsafe option. It means that when no source root can be "
+        "found for a file, such as `src/python` or `tests/java`, Pants will try to automatically "
+        "create an implicit source root. Instead, when Pants does not already know about the "
```

Comment:
Rather than suggesting that folks do this, would be better to suggest that they consider organizing their code around the built in patterns...? IE, let's be more opinionated here.

We should also ensure that the message you get for a `fail` is good.

Reply:
>  pants.source.source_root.NoSourceRootError: Could not find a source root for `build-support/migration-support/BUILD`.